File: Hard/C++/Maximal_Rectangle(85).cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if (matrix.empty()) { return 0; }
		int length = matrix[0].size();
		int res = 0;
		vector<int> current_row(length+1);
		current_row[length] = -1;
		for (int i = 0; i < matrix.size(); i++) {
			
			for (int j = 0; j < length; j++) {
				
				if (matrix[i][j] == '1') { current_row[j]++; }
				else { current_row[j] = 0; }
			}
			
			stack<int> st;
			current_row[length] = -1;
			
			for (int m = 0; m <= length; m++) {
				while (!st.empty() && current_row[st.top()] > current_row[m]) {
					int top=st.top();
					st.pop();
					int bot = 0;
					if (!st.empty()) {
						bot = m - st.top() -1;
					}
					else {
						bot = m;
					}
					res = (bot * current_row[top]) > res ? (bot * current_row[top]) : res;
				}
				st.push(m);
			}
		}
		return res;
    }
};

File: Hard/Java/Maximal_Rectangle(85).java
class Solution {
    
    public int maximalRectangle(char[][] matrix) {
        
        if (matrix.length==0) return 0;
        
        int m = matrix.length, n = matrix[0].length, result = 0;
        int[][] dp = new int[m][n];
        
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (matrix[i][j]=='1')
                    dp[i][j] = i==0 ? 1 : dp[i-1][j] + 1;
            }
        }
        
        for (int i=0; i<m; i++)
            result = Math.max(result, largestRectangleArea(dp[i]));
        
        return result;
    }
    
    private int largestRectangleArea(int[] heights) {
        
        Stack<Integer> stack = new Stack<>();
        int n = heights.length, i=0, res = 0;
        
        while (i<n || !stack.isEmpty()) {
            
            while (!stack.isEmpty() && (i==n || heights[i]<heights[stack.peek()])) {
                int topElementIndex = stack.pop();
                int startIndex = stack.size()==0 ? 0 : stack.peek() + 1;
                int area = heights[topElementIndex] * (i-startIndex);
                res = Math.max(res, area);
            }
            
            if (i<n) stack.push(i);
            i++;
        }
        
        return res;
    }
}